File: android/app/src/main/kotlin/app/zerorelay/data/crypto/RelayCrypto.kt
package app.zerorelay.data.crypto

import android.util.Base64
import org.bouncycastle.crypto.generators.HKDFBytesGenerator
import org.bouncycastle.crypto.params.Ed25519PrivateKeyParameters
import org.bouncycastle.crypto.params.Ed25519PublicKeyParameters
import org.bouncycastle.crypto.signers.Ed25519Signer
import org.bouncycastle.crypto.digests.SHA256Digest
import org.bouncycastle.crypto.params.HKDFParameters
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * 中继层：房间访问令牌 + Ed25519 消息签名（与 server/relay-security.ts、cli.js 对齐）
 */
object RelayCrypto {
    private const val ROOM_ACCESS_INFO = "zero-relay-room-access-v1"
    private const val SIGN_SEED_INFO = "ed25519-seed"
    private const val SIGN_HKDF_SALT = "zero-relay-sign-v1"

    data class SigningKeys(
        val signPrivate: ByteArray,
        val signPublic: ByteArray,
    )

    fun routeHashFromToken(tokenB64: String): ByteArray? {
        val raw = try {
            Base64.decode(tokenB64, Base64.NO_WRAP)
        } catch (_: Exception) {
            return null
        }
        if (raw.size != 32) return null
        return java.security.MessageDigest.getInstance("SHA-256").digest(raw)
    }

    fun routeHashB64FromToken(tokenB64: String): String? {
        val hash = routeHashFromToken(tokenB64) ?: return null
        return Base64.encodeToString(hash, Base64.NO_WRAP)
    }

    fun routeIdFromToken(tokenB64: String): String? {
        val hash = routeHashFromToken(tokenB64) ?: return null
        return Base64.encodeToString(hash, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)
    }

    fun deriveRoomAccessToken(roomSecret: ByteArray, roomId: String): String {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(roomSecret, "HmacSHA256"))
        mac.update(ROOM_ACCESS_INFO.toByteArray(Charsets.UTF_8))
        mac.update(roomId.toByteArray(Charsets.UTF_8))
        return Base64.encodeToString(mac.doFinal(), Base64.NO_WRAP)
    }

    fun deriveSigningKeys(identityPrivateKey: ByteArray): SigningKeys {
        val seed = hkdf(identityPrivateKey, SIGN_HKDF_SALT.toByteArray(), SIGN_SEED_INFO.toByteArray())
        val privateParams = Ed25519PrivateKeyParameters(seed, 0)
        val publicParams = privateParams.generatePublicKey()
        return SigningKeys(
            signPrivate = privateParams.encoded,
            signPublic = publicParams.encoded,
        )
    }

    fun encodeSignPublicKey(signPublic: ByteArray): String =
        Base64.encodeToString(signPublic, Base64.NO_WRAP)

    fun buildSignPayload(
        routeId: String,
        senderId: String,
        timestamp: Long,
        ciphertext: String,
        iv: String,
        tag: String,
    ): ByteArray {
        val line = listOf(routeId, senderId, timestamp.toString(), ciphertext, iv, tag).joinToString("\n")
        return line.toByteArray(Charsets.UTF_8)
    }

    fun signMessage(
        signingKeys: SigningKeys,
        routeId: String,
        senderId: String,
        timestamp: Long,
        ciphertext: String,
        iv: String,
        tag: String,
    ): String {
        val payload = buildSignPayload(routeId, senderId, timestamp, ciphertext, iv, tag)
        val signer = Ed25519Signer()
        signer.init(true, Ed25519PrivateKeyParameters(signingKeys.signPrivate, 0))
        signer.update(payload, 0, payload.size)
        return Base64.encodeToString(signer.generateSignature(), Base64.NO_WRAP)
    }

    fun verifyMessage(
        senderPk: ByteArray,
        signPk: ByteArray,
        sig: ByteArray,
        routeId: String,
        senderId: String,
        timestamp: Long,
        ciphertext: String,
        iv: String,
        tag: String,
    ): Boolean {
        if (IdentityCrypto.senderIdFromPublicKey(senderPk) != senderId) return false
        val payload = buildSignPayload(routeId, senderId, timestamp, ciphertext, iv, tag)
        val verifier = Ed25519Signer()
        verifier.init(false, Ed25519PublicKeyParameters(signPk, 0))
        verifier.update(payload, 0, payload.size)
        return verifier.verifySignature(sig)
    }

    /** 隧道 AES 密钥：HKDF(SHA256(roomToken))，与帧内 routeHash 一致 */
    fun tunnelKeyFromRouteHash(routeHash: ByteArray): ByteArray {
        require(routeHash.size == 32)
        return hkdf(routeHash, ByteArray(0), "zero-relay-tunnel-v1".toByteArray(Charsets.UTF_8))
    }

    private fun hkdf(ikm: ByteArray, salt: ByteArray, info: ByteArray): ByteArray {
        val gen = HKDFBytesGenerator(SHA256Digest())
        gen.init(HKDFParameters(ikm, salt, info))
        val out = ByteArray(32)
        gen.generateBytes(out, 0, out.size)
        return out
    }
}

File: web/src/crypto/relay-crypto.ts
import { ed25519 } from '@noble/curves/ed25519';
import { b64decode, b64encode, b64urlEncode, sha256, utf8 } from './bytes';
import { hkdfSha256 } from './hkdf';

const ROOM_ACCESS_INFO = 'zero-relay-room-access-v1';
const SIGN_HKDF_SALT = 'zero-relay-sign-v1';
const SIGN_SEED_INFO = 'ed25519-seed';

export interface SigningKeys {
  signPrivate: Uint8Array;
  signPublic: string;
}

export interface RelayMessage {
  id?: string;
  senderId: string;
  senderPk?: string;
  signPk?: string;
  sig?: string;
  ciphertext: string;
  iv: string;
  tag: string;
  timestamp: number;
}

const senderIdCache = new Map<string, string>();

export async function deriveRoomAccessToken(roomSecret: Uint8Array, roomId: string): Promise<string> {
  const key = await crypto.subtle.importKey(
    'raw',
    roomSecret,
    { name: 'HMAC', hash: 'SHA-256' },
    false,
    ['sign'],
  );
  const msg = utf8(ROOM_ACCESS_INFO + roomId);
  const sig = await crypto.subtle.sign('HMAC', key, msg);
  return b64encode(new Uint8Array(sig));
}

export async function routeHashFromToken(tokenB64: string): Promise<Uint8Array | null> {
  try {
    const raw = b64decode(tokenB64);
    if (raw.length !== 32) return null;
    return sha256(raw);
  } catch {
    return null;
  }
}

export async function routeIdFromToken(tokenB64: string): Promise<string | null> {
  const hash = await routeHashFromToken(tokenB64);
  return hash ? b64urlEncode(hash) : null;
}

export async function deriveSigningKeys(identityPrivateKeyB64: string): Promise<SigningKeys> {
  const seed = await hkdfSha256(
    b64decode(identityPrivateKeyB64),
    utf8(SIGN_HKDF_SALT),
    SIGN_SEED_INFO,
  );
  return { signPrivate: seed, signPublic: b64encode(ed25519.getPublicKey(seed)) };
}

function buildSignPayload(
  routeId: string,
  senderId: string,
  timestamp: number,
  ciphertext: string,
  iv: string,
  tag: string,
): Uint8Array {
  return utf8([routeId, senderId, String(timestamp), ciphertext, iv, tag].join('\n'));
}

export function signMessage(
  keys: SigningKeys,
  routeId: string,
  senderId: string,
  timestamp: number,
  ciphertext: string,
  iv: string,
  tag: string,
): string {
  return b64encode(
    ed25519.sign(buildSignPayload(routeId, senderId, timestamp, ciphertext, iv, tag), keys.signPrivate),
  );
}

export async function senderIdFromPublicKey(pubB64: string): Promise<string> {
  const cached = senderIdCache.get(pubB64);
  if (cached) return cached;
  const hash = await sha256(b64decode(pubB64));
  const id =
    'id_' +
    b64urlEncode(hash.subarray(0, 6))
      .replace(/-/g, 'x')
      .replace(/_/g, 'y');
  senderIdCache.set(pubB64, id);
  return id;
}

export function verifyMessage(msg: RelayMessage & { routeId: string }): boolean {
  const { senderPk, signPk, sig, senderId, routeId, ciphertext, iv, tag, timestamp } = msg;
  if (!senderPk || !signPk || !sig) return false;
  const cached = senderIdCache.get(senderPk);
  if (cached && cached !== senderId) return false;
  const payload = buildSignPayload(routeId, senderId, timestamp, ciphertext, iv, tag);
  try {
    return ed25519.verify(b64decode(sig), payload, b64decode(signPk));
  } catch {
    return false;
  }
}
